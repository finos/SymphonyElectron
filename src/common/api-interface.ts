export enum apiCmds {
    isOnline = 'is-online',
    getVersionInfo = 'get-version-info',
    registerLogger = 'register-logger',
    setBadgeCount = 'set-badge-count',
    badgeDataUrl = 'badge-data-url',
    activate = 'activate',
    registerBoundsChange = 'register-bounds-change',
    registerProtocolHandler = 'register-protocol-handler',
    registerLogRetriever = 'register-log-retriever',
    sendLogs = 'send-logs',
    registerAnalyticsHandler = 'register-analytics-handler',
    registerActivityDetection = 'register-activity-detection',
    showNotificationSettings = 'show-notification-settings',
    sanitize = 'sanitize',
    bringToFront = 'bring-to-front',
    openScreenPickerWindow = 'open-screen-picker-window',
    popupMenu = 'popup-menu',
    optimizeMemoryConsumption = 'optimize-memory-consumption',
    optimizeMemoryRegister = 'optimize-memory-register',
    setIsInMeeting = 'set-is-in-meeting',
    setLocale = 'set-locale',
    openScreenSnippet = 'open-screen-snippet',
    closeScreenSnippet = 'close-screen-snippet',
    keyPress = 'key-press',
    closeWindow = 'close-window',
    openScreenSharingIndicator = 'open-screen-sharing-indicator',
    closeScreenSharingIndicator = 'close-screen-sharing-indicator',
    downloadManagerAction = 'download-manager-action',
    getMediaSource = 'get-media-source',
    notification = 'notification',
    closeNotification = 'close-notification',
    isMisspelled = 'is-misspelled',
    memoryInfo = 'memory-info',
    swiftSearch = 'swift-search',
    getConfigUrl = 'get-config-url',
    registerRestartFloater = 'register-restart-floater',
    setCloudConfig = 'set-cloud-config',
    getCPUUsage = 'get-cpu-usage',
    checkMediaPermission = 'check-media-permission',
    registerDownloadHandler = 'register-download-handler',
    openDownloadedItem = 'open-downloaded-item',
    showDownloadedItem = 'show-downloaded-item',
    clearDownloadedItems = 'clear-downloaded-items',
}

export enum apiName {
    symphonyApi = 'symphony-api',
    mainWindowName = 'main',
    notificationWindowName = 'notification-window',
}

export interface IApiArgs {
    memoryInfo: Electron.ProcessMemoryInfo;
    word: string;
    cmd: apiCmds;
    isOnline: boolean;
    count: number;
    dataUrl: string;
    windowName: string;
    period: number;
    reason: string;
    sources: Electron.DesktopCapturerSource[];
    id: number;
    cpuUsage: Electron.CPUUsage;
    isInMeeting: boolean;
    locale: string;
    keyCode: number;
    windowType: WindowTypes;
    winKey: string;
    streamId: string;
    displayId: string;
    path: string;
    type: string;
    logName: string;
    logs: ILogs;
    cloudConfig: object;
}

export type WindowTypes = 'screen-picker' | 'screen-sharing-indicator' | 'notification-settings';

export interface IBadgeCount {
    count: number;
}

/**
 * Screen snippet
 */
export type ScreenSnippetDataType = 'ERROR' | 'image/png;base64';
export interface IScreenSnippet {
    data?: string;
    message?: string;
    type: ScreenSnippetDataType;
}

export interface IBoundsChange extends Electron.Rectangle {
    windowName: string;
}

/**
 * Screen sharing indicator
 */
export interface IScreenSharingIndicator {
    type: string;
    requestId: number;
    reason?: string;
}

export enum KeyCodes {
    Esc = 27,
    Alt = 18,
}

type Theme = '' | 'light' | 'dark';

/**
 * Notification
 */
export interface INotificationData {
    id: number;
    title: string;
    body: string;
    image: string;
    icon: string;
    flash: boolean;
    color: string;
    tag: string;
    sticky: boolean;
    company: string;
    displayTime: number;
    isExternal: boolean;
    theme: Theme;
}

export enum NotificationActions {
    notificationClicked = 'notification-clicked',
    notificationClosed = 'notification-closed',
}

/**
 * Screen sharing Indicator
 */
export interface IScreenSharingIndicatorOptions {
    displayId: string;
    requestId: number;
    streamId: string;
    stream?: MediaStream;
}

export interface IVersionInfo {
    containerIdentifier: string;
    containerVer: string;
    buildNumber: string;
    apiVer: string;
    searchApiVer: string;
}

export interface ICPUUsage {
    percentCPUUsage: number;
    idleWakeupsPerSecond: number;
}

export interface IDownloadManager {
    _id: string;
    fileName: string;
    fileDisplayName: string;
    savedPath: string;
    total: number;
    flashing?: boolean;
    count?: number;
}

export interface IMediaPermission {
    camera: string;
    microphone: string;
    screen: string;
}

export interface ILogMsg {
    level: LogLevel;
    details: any;
    showInConsole: boolean;
    startTime: number;
}

export type LogLevel = 'error' | 'warn' | 'info' | 'verbose' | 'debug' | 'silly';

export interface ILogFile {
    filename: string;
    contents: string;
}

export interface ILogs {
    logName: string;
    logFiles: ILogFile[];
}

export interface IRestartFloaterData {
    windowName: string;
    bounds: Electron.Rectangle;
}
