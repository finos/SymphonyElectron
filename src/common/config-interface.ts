import { IConfig } from '../app/config-handler';

export const ConfigFieldsDefaultValues: Partial<IConfig> = {
  isPodUrlEditable: true,
  forceAutoUpdate: false,
  enableBrowserLogin: false,
  browserLoginAutoConnect: false,
  latestAutoUpdateChannelEnabled: true,
  betaAutoUpdateChannelEnabled: true,
  browserLoginRetryTimeout: '5',
  openfin: {
    uuid: '',
    licenseKey: '',
    runtimeVersion: '',
    platformUuid: '',
    autoConnect: false,
    connectionTimeout: '10000',
  },
};
