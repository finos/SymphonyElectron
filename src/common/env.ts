export const isDevEnv = process.env.ELECTRON_DEV
  ? process.env.ELECTRON_DEV.trim().toLowerCase() === 'true'
  : false;
export const isElectronQA = !!process.env.ELECTRON_QA;

export const isMac = process.platform === 'darwin';
export const isWindowsOS = process.platform === 'win32';
export const isLinux = process.platform === 'linux';
