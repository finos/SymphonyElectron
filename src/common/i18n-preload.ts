import { formatString } from './utils';

const localeCodeRegex = /^([a-z]{2})-([A-Z]{2})$/;

export type LocaleType = 'en-US' | 'ja-JP';

type formatterFunction = (args?: object) => string;

class Translation {

    /**
     * Returns translated string with respect to value, resource & name space
     *
     * @param value {string} key field in the resources
     * @param resource {string} current locale resource
     * @param namespace {string} name space in the resource
     */
    private static translate(value: string, resource: JSON | null, namespace: string | undefined): string {
        return resource ? (Translation.getResource(resource, namespace)[value] || value) : value;
    }
    private static getResource = (resource: JSON, namespace: string | undefined): JSON => namespace ? resource[namespace] : resource;
    private locale: LocaleType = 'en-US';
    private loadedResources: object = {};

    /**
     * Apply the locale for translation
     *
     * @param locale
     */
    public setLocale(locale: LocaleType): void {
        const localeMatch: string[] | null = locale.match(localeCodeRegex);
        if (!locale && (!localeMatch || localeMatch.length < 1)) {
            return;
        }

        this.locale = locale;
    }

    /**
     * Gets the current locale
     */
    public getLocale(): LocaleType {
        return this.locale;
    }

    /**
     * fetches and returns the translated value
     *
     * @param value {string}
     * @param namespace {string}
     * @example t('translate and formats {data} ', namespace)({ data: 'string' })
     * @returns translate and formats string
     */
    public t(value: string, namespace?: string): formatterFunction {
        return (args?: object): string => {
            if (this.loadedResources && this.loadedResources[this.locale]) {
                return formatString(Translation.translate(value, this.loadedResources[this.locale], namespace), args);
            }
            const resource = this.loadResource(this.locale);
            return formatString(Translation.translate(value, resource, namespace) || value, args);
        };
    }

    /**
     * Keeps ref of loaded resources from the main process
     *
     * @param locale {LocaleType}
     * @param resource {JSON}
     */
    public setResource(locale: LocaleType, resource: JSON): void {
        this.locale = locale;
        this.loadedResources = resource;
    }

    /**
     * Reads the resources dir and returns the data
     *
     * @param locale
     */
    private loadResource(locale: LocaleType): JSON | null {
        return this.loadedResources[locale];
    }

}

const i18n = new Translation();

export { i18n };
