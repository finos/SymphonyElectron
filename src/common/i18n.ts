import * as fs from 'fs';
import * as path from 'path';

import { formatString } from './utils';

const localeCodeRegex = /^([a-z]{2})-([A-Z]{2})$/;

export type LocaleType = 'en-US' | 'ja-JP';

type formatterFunction = (args?: object) => string;

class Translation {
    /**
     * Returns translated string with respect to value, resource & name space
     *
     * @param value {string} key field in the resources
     * @param resource {string} current locale resource
     * @param namespace {string} name space in the resource
     */
    private static translate(value: string, resource: JSON | null, namespace: string | undefined): string {
        return resource ? (Translation.getResource(resource, namespace)[value] || value) : value;
    }
    private static getResource = (resource: JSON, namespace: string | undefined): JSON => namespace ? resource[namespace] : resource;
    public loadedResources: object = {};
    private locale: LocaleType = 'en-US';

    /**
     * Apply the locale for translation
     *
     * @param locale
     */
    public setLocale(locale: LocaleType): void {
        const localeMatch: string[] | null = locale.match(localeCodeRegex);
        if (!locale && (!localeMatch || localeMatch.length < 1)) {
            return;
        }

        this.locale = locale;
    }

    /**
     * Gets the current locale
     */
    public getLocale(): LocaleType {
        return this.locale;
    }

    /**
     * fetches and returns the translated value
     *
     * @param value {string}
     * @param namespace {string}
     * @example t('translate and formats {data} ', namespace)({ data: 'string' })
     * @returns translate and formats string
     */
    public t(value: string, namespace?: string): formatterFunction {
        return (args?: object): string => {
            if (this.loadedResources && this.loadedResources[this.locale]) {
                return formatString(Translation.translate(value, this.loadedResources[this.locale], namespace), args);
            }
            const resource = this.loadResource(this.locale);
            return formatString(Translation.translate(value, resource, namespace) || value, args);
        };
    }

    /**
     * Reads the resources dir and returns the data
     *
     * @param locale
     */
    private loadResource(locale: LocaleType): JSON | null {
        const resourcePath = path.resolve(__dirname, '..', 'locale', `${locale}.json`);

        if (!fs.existsSync(resourcePath)) {
            return null;
        }

        return this.loadedResources[this.locale] = require(resourcePath);
    }

}

const i18n = new Translation();

export { i18n };
