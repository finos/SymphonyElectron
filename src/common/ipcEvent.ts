export enum ScreenShotAnnotation {
  COPY_TO_CLIPBOARD = 'copy-to-clipboard',
  SAVE_AS = 'save-as',
  CLOSE = 'close-snippet',
  UPLOAD = 'upload-snippet',
}
