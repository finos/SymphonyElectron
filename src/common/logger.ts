import { app } from 'electron';
import electronLog, { LogLevel, transports } from 'electron-log';
import * as fs from 'fs';
import * as path from 'path';
import * as util from 'util';

import { isElectronQA } from './env';
import { getCommandLineArgs } from './utils';

export interface ILogMsg {
    level: LogLevel;
    details: any;
    showInConsole: boolean;
    startTime: number;
}

interface IClientLogMsg {
    msgs?: ILogMsg[];
    logLevel?: LogLevel;
    showInConsole?: boolean;
}

const MAX_LOG_QUEUE_LENGTH = 100;

class Logger {
    private readonly showInConsole: boolean = false;
    private readonly desiredLogLevel?: LogLevel;
    private readonly logQueue: ILogMsg[];
    private readonly logPath: string;
    private loggerWindow: Electron.WebContents | null;

    constructor() {

        this.loggerWindow = null;
        this.logQueue = [];
        // If the user has specified a custom log path use it.
        const customLogPathArg = getCommandLineArgs(process.argv, '--logPath=', false);
        const customLogsFolder = customLogPathArg && customLogPathArg.substring(customLogPathArg.indexOf('=') + 1);
        if (customLogsFolder && fs.existsSync(customLogsFolder)) {
            app.setPath('logs', customLogsFolder);
        }

        this.logPath = app.getPath('logs');

        if (!isElectronQA) {
            transports.file.file = path.join(this.logPath, `app_${Date.now()}.log`);
            transports.file.level = 'debug';
            transports.file.format = '{y}-{m}-{d} {h}:{i}:{s}:{ms} {z} | {level} | {text}';
            transports.file.appName = 'Symphony';
        }

        const logLevel = getCommandLineArgs(process.argv, '--logLevel=', false);
        if (logLevel) {
            const level = logLevel.split('=')[1];
            if (level) {
                this.desiredLogLevel = level as LogLevel;
            }
        }

        if (getCommandLineArgs(process.argv, '--enableConsoleLogging', false)) {
            this.showInConsole = true;
        }

        // cleans up old logs if there are any
        this.cleanupOldLogs();
    }

    /**
     * get instance of logQueue
     */
    public getLogQueue(): ILogMsg[] {
        return this.logQueue;
    }

    /**
     * Log error
     *
     * @param message {string} - message to be logged
     * @param data {any} - extra data that needs to be logged
     */
    public error(message: string, ...data: any[]): void {
        this.log('error', message, data);
    }

    /**
     * Log warn
     *
     * @param message {string} - message to be logged
     * @param data {any} - extra data that needs to be logged
     */
    public warn(message: string, ...data: any[]): void {
        this.log('warn', message, data);
    }

    /**
     * Log info
     *
     * @param message {string} - message to be logged
     * @param data {any} - extra data that needs to be logged
     */
    public info(message: string, ...data: any[]): void {
        this.log('info', message, data);
    }

    /**
     * Log verbose
     *
     * @param message {string} - message to be logged
     * @param data {array} - extra data that needs to be logged
     */
    public verbose(message: string, ...data: any[]): void {
        this.log('verbose', message, data);
    }

    /**
     * Log debug
     *
     * @param message {string} - message to be logged
     * @param data {any} - extra data that needs to be logged
     */
    public debug(message: string, ...data: any[]): void {
        this.log('debug', message, data);
    }

    /**
     * Log silly
     *
     * @param message {string} - message to be logged
     * @param data {any} - extra data that needs to be logged
     */
    public silly(message: string, ...data: any[]): void {
        this.log('silly', message, data);
    }

    /**
     * Sets the renderer window for sending logs to the client
     *
     * @param window {WebContents} - renderer window
     */
    public setLoggerWindow(window: Electron.WebContents): void {
        this.loggerWindow = window;

        if (this.loggerWindow) {
            const logMsgs: IClientLogMsg = {};
            if (this.logQueue.length) {
                logMsgs.msgs = this.logQueue;
            }
            if (this.desiredLogLevel) {
                logMsgs.logLevel = this.desiredLogLevel;
            }
            if (Object.keys(logMsgs).length) {
                this.loggerWindow.send('log', logMsgs);
            }
        }
    }

    /**
     * Main instance of the logger method
     *
     * @param logLevel {LogLevel} - Different type of log levels
     * @param message {string} - Log message
     * @param data {array} - extra data to be logged
     */
    private log(logLevel: LogLevel, message: string, data: any[] = []): void {
        if (data && data.length > 0) {
            data.forEach((param) => {
                message += `, '${param && typeof param}': ${JSON.stringify(param)}`;
            });
        }
        if (!isElectronQA) {
            switch (logLevel) {
                case 'error': electronLog.error(message); break;
                case 'warn': electronLog.warn(message); break;
                case 'info': electronLog.info(message); break;
                case 'verbose': electronLog.verbose(message); break;
                case 'debug': electronLog.debug(message); break;
                case 'silly': electronLog.silly(message); break;
                default: electronLog.info(message);
            }
        }
        this.sendToCloud(this.formatLogMsg(logLevel, message));
    }

    /**
     * Formats the logs in the format that required
     * to send to the client
     *
     * @param level {LogLevel} - Different type of log levels
     * @param details {any} - log format that required to send to client
     */
    private formatLogMsg(level: LogLevel, details: any): ILogMsg {
        return {
            details,
            level,
            showInConsole: this.showInConsole,
            startTime: Date.now(),
        };
    }

    /**
     * This will send the logs to the client if loggerWindow
     * else adds the logs to a Queue
     *
     * @param logMsg {ILogMsg}
     */
    private sendToCloud(logMsg: ILogMsg): void {
        // don't send logs if it is not desired by the user
        if (this.desiredLogLevel && this.desiredLogLevel !== logMsg.level) {
            return;
        }

        if (this.loggerWindow) {
            this.loggerWindow.send('log', { msgs: [ logMsg ], logLevel: this.desiredLogLevel, showInConsole: this.showInConsole });
            return;
        }

        this.logQueue.push(logMsg);
        // don't store more than 100 msgs. keep most recent log msgs.
        if (this.logQueue.length > MAX_LOG_QUEUE_LENGTH) {
            this.logQueue.shift();
        }
    }

    /**
     * Cleans up logs older than a day
     */
    private cleanupOldLogs(): void {
        const files = fs.readdirSync(this.logPath);
        const deleteTimeStamp = new Date().getTime() + (10 * 24 * 60 * 60 * 1000);
        files.forEach((file) => {
            if (file === '.DS_Store' || file === 'app.log') {
                return;
            }
            const filePath = path.join(this.logPath, file);
            const stat = fs.statSync(filePath);
            const fileTimestamp = new Date(util.inspect(stat.mtime)).getTime();
            if (fileTimestamp > deleteTimeStamp) {
                fs.unlinkSync(filePath);
            }
        });
    }
}

const logger = new Logger();

export { logger };
