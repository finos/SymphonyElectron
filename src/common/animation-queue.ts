import { logger } from './logger';

export class AnimationQueue {
  private queue: any[] = [];
  private running: boolean = false;

  constructor() {
    this.animate = this.animate.bind(this);
  }

  /**
   * Pushes each animation to a queue
   *
   * @param object
   */
  public push(object): void {
    if (this.running) {
      this.queue.push(object);
      return;
    }
    this.running = true;
    setTimeout(() => this.animate(object), 0);
  }

  /**
   * Animates an animation that is part of the queue
   * @param object
   */
  public async animate(object): Promise<void> {
    try {
      await object.func.apply(null, object.args);
    } catch (err: any) {
      logger.error(
        `animation-queue: encountered an error: ${err} with stack trace: ${err.stack}`,
      );
    } finally {
      if (this.queue.length > 0) {
        // Run next animation
        this.animate.call(this, this.queue.shift());
      } else {
        this.running = false;
      }
    }
  }

  /**
   * Clears the queue
   */
  public clear(): void {
    this.queue = [];
  }
}
