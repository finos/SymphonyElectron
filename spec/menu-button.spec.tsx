import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { ScreenShotAnnotation } from '../src/common/ipcEvent';
import MenuButton from '../src/renderer/components/menu-button';

afterEach(() => {
  jest.clearAllMocks();
});

describe('Menu Button', () => {
  //#region Logic
  const waitForPromisesToResolve = () =>
    new Promise((resolve) => setTimeout(resolve));
  //#endregion

  //#region Mock
  const copyToClipboardFn = jest.fn();
  const saveAsFn = jest.fn();
  copyToClipboardFn.mockImplementation((event) => {
    return event;
  });
  saveAsFn.mockImplementation((event) => {
    return event;
  });
  const menuItem = [
    {
      name: 'Copy to clipboard',
      event: ScreenShotAnnotation.COPY_TO_CLIPBOARD,
      onClick: copyToClipboardFn,
      dataTestId: 'COPY_TO_CLIPBOARD',
    },
    {
      name: 'Save as',
      event: ScreenShotAnnotation.SAVE_AS,
      onClick: saveAsFn,
      dataTestId: 'SAVE_AS',
    },
  ];
  //#endregion

  it('should show all elements', () => {
    const wrapper = shallow(React.createElement(MenuButton));
    expect(wrapper).toMatchSnapshot();
  });

  it('should call event on click copy to clipboard', async () => {
    const wrapper = mount(
      <MenuButton id='snipping-tool' listItems={menuItem} />,
    );

    wrapper.find('[data-testid="snipping-tool_MENU_BUTTON"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    wrapper
      .find('[data-testid="snipping-tool_COPY_TO_CLIPBOARD"]')
      .simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    expect(copyToClipboardFn).toBeCalledWith(
      ScreenShotAnnotation.COPY_TO_CLIPBOARD,
    );
  });

  it('should call event on click save as', async () => {
    const wrapper = mount(
      <MenuButton id='snipping-tool' listItems={menuItem} />,
    );

    wrapper.find('[data-testid="snipping-tool_MENU_BUTTON"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    wrapper.find('[data-testid="snipping-tool_SAVE_AS"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    expect(saveAsFn).toBeCalledWith(ScreenShotAnnotation.SAVE_AS);
  });

  it('should disappear on click to other element', async () => {
    const wrapper = mount(
      <MenuButton id='snipping-tool' listItems={menuItem} />,
    );

    wrapper.find('[data-testid="snipping-tool_MENU_BUTTON"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    wrapper.find('[data-testid="snipping-tool_MENU_BUTTON"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    expect(wrapper.find('.menu').exists()).toBeFalsy();
  });

  it('should focus menu list on up', async () => {
    const wrapper = mount(
      <MenuButton id='snipping-tool' listItems={menuItem} />,
    );

    wrapper.find('[data-testid="snipping-tool_MENU_BUTTON"]').simulate('click');
    wrapper.update();
    wrapper
      .find('[data-testid="snipping-tool_MENU_BUTTON"]')
      .simulate('keyup', { key: 'ArrowUp' });
    wrapper.update();
    expect(
      (wrapper.find('[data-testid="snipping-tool_SAVE_AS"]').instance() as any)
        .dataset.isfocused,
    ).toBe('true');
  });

  it('should focus menu list on down', async () => {
    const wrapper = mount(
      <MenuButton id='snipping-tool' listItems={menuItem} />,
    );

    wrapper.find('[data-testid="snipping-tool_MENU_BUTTON"]').simulate('click');
    wrapper.update();
    wrapper
      .find('[data-testid="snipping-tool_MENU_BUTTON"]')
      .simulate('keyup', { key: 'ArrowDown' });
    wrapper.update();
    expect(
      (
        wrapper
          .find('[data-testid="snipping-tool_COPY_TO_CLIPBOARD"]')
          .instance() as any
      ).dataset.isfocused,
    ).toBe('true');
  });

  it('should go down on press down', async () => {
    const wrapper = mount(
      <MenuButton id='snipping-tool' listItems={menuItem} />,
    );

    wrapper.find('[data-testid="snipping-tool_MENU_BUTTON"]').simulate('click');
    wrapper.update();
    wrapper
      .find('[data-testid="snipping-tool_MENU_BUTTON"]')
      .simulate('keyup', { key: 'ArrowDown' });
    wrapper.update();
    wrapper
      .find('[data-testid="snipping-tool_LIST"]')
      .simulate('keyup', { key: 'ArrowDown' });
    wrapper.update();
    expect(
      (wrapper.find('[data-testid="snipping-tool_SAVE_AS"]').instance() as any)
        .dataset.isfocused,
    ).toBe('true');
  });

  it('should go up on press up', async () => {
    const wrapper = mount(
      <MenuButton id='snipping-tool' listItems={menuItem} />,
    );

    wrapper.find('[data-testid="snipping-tool_MENU_BUTTON"]').simulate('click');
    wrapper.update();
    wrapper
      .find('[data-testid="snipping-tool_MENU_BUTTON"]')
      .simulate('keyup', { key: 'ArrowUp' });
    wrapper
      .find('[data-testid="snipping-tool_LIST"]')
      .simulate('keyup', { key: 'ArrowUp' });
    wrapper.update();
    expect(
      (
        wrapper
          .find('[data-testid="snipping-tool_COPY_TO_CLIPBOARD"]')
          .instance() as any
      ).dataset.isfocused,
    ).toBe('true');
  });
});
