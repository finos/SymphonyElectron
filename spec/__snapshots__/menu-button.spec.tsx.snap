// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Button should show all elements 1`] = `
<Fragment>
  <div
    className="menu-button-wrapper"
  >
    <button
      className="menu-button"
      data-testid="undefined_MENU_BUTTON"
      onClick={[Function]}
    >
      <img
        alt="Open menu"
        draggable={false}
        src="../renderer/assets/single-chevron-down.svg"
        title="Open menu"
      />
    </button>
  </div>
</Fragment>
`;
